File: database/addUser.py
# Copyright 2017 Panel Spy.  All rights reserved.

import printctl
import argparse
import json
import context

printctl.off()
import sql

if __name__ == '__main__':
    parser = argparse.ArgumentParser( description='add user' )
    parser.add_argument( '-b', '--by', dest='by', help='requested by username' )
    parser.add_argument( '-u', '--username', dest='username', help='username' )
    parser.add_argument( '-p', '--password', dest='password', help='password' )
    parser.add_argument( '-r', '--role', dest='role', help='role' )
    parser.add_argument( '-s', '--status', dest='status', help='role' )
    parser.add_argument( '-f', '--first_name', dest='first_name', help='role' )
    parser.add_argument( '-l', '--last_name', dest='last_name', help='role' )
    parser.add_argument( '-e', '--email_address', dest='email_address', help='role' )
    parser.add_argument( '-g', '--organization', dest='organization', help='role' )
    parser.add_argument( '-d', '--description', dest='description', help='description' )
    parser = context.add_context_args( parser )
    args = parser.parse_args()

    try:
        status = sql.addUser( args.by, args.username, args.password, args.role, args.status, args.first_name, args.last_name, args.email_address, args.organization, args.description, args.enterprise )
    except:
        dict = { 'Error': 'Failed to add user' }
    else:
        dict = status.__dict__

    printctl.on()
    print( json.dumps( dict ) )

File: users/addUser.php
<?php
  // Copyright 2018 Panel Spy.  All rights reserved.
  require_once $_SERVER["DOCUMENT_ROOT"] . "/../common/util.php";
  require_once $_SERVER["DOCUMENT_ROOT"]."/util/postSecurity.php";
  require_once $_SERVER["DOCUMENT_ROOT"]."/util/context.php";

  // Log post array without password
  $sPasswordMask = '<PASSWORD>';
  $aPostMinusPassword = $_POST;
  $aPostMinusPassword['password'] = $sPasswordMask;
  error_log( "==> post=" . print_r( $aPostMinusPassword, true ) );

  // Get user attributes
  $sUsername = $_POST['username'];
  $sPassword = quote( $_POST['password'], false );
  $sRole = $_POST['role'];
  $sAuthFacilities = quote( $_POST['auth_facilities'] );
  $sStatus = $_POST['status'];
  $sFirstName = quote( $_POST['first_name'] );
  $sLastName = quote( $_POST['last_name'] );
  $sEmailAddress = quote( $_POST['email_address'] );
  $sOrganization = quote( $_POST['organization'] );
  $sDescription = quote( $_POST['user_description'] );

  // Add user
  $command = quote( getenv( "PYTHON" ) ) . " ../database/addUser.py 2>&1 -b " . $_SESSION['panelSpy']['user']['username']
    . ' -u ' . $sUsername
    . ' -p ' . $sPassword
    . ' -r ' . $sRole
    . ' -a ' . $sAuthFacilities
    . ' -s ' . $sStatus
    . ' -f ' . $sFirstName
    . ' -l ' . $sLastName
    . ' -e ' . $sEmailAddress
    . ' -g ' . $sOrganization
    . ' -d ' . $sDescription
    . $g_sContext;

  error_log( "==> command=" . preg_replace( '/ -p .* -r/', ' -p ' . $sPasswordMask . ' -r', $command, 1 ) );
  exec( $command, $output, $status );
  error_log( "==> output=" . print_r( $output, true ) );

  // Echo status
  $sStatus = $output[ count( $output ) - 1 ];
  echo $sStatus;
?>
